File: backend/app/models/order.py
from typing import Optional, List, Dict
from beanie import Document
from pydantic import Field
from datetime import datetime

class Order(Document):
    """Order model for storing purchase information."""
    user: str  # User ID
    items: List[Dict] = Field(default_factory=list)  # List of ordered items with product details
    totalPrice: float = 0.0  # Total order amount
    shippingAddress: Dict = Field(default_factory=dict)  # Shipping details
    orderStatus: str = "pending"  # pending, confirmed, shipped, delivered, cancelled
    paymentMethod: str = "razorpay"  # Payment method
    isPaid: bool = False  # Payment status
    paymentId: Optional[str] = None  # Razorpay payment ID
    paidAt: Optional[datetime] = None  # Payment timestamp
    isDelivered: bool = False  # Delivery status
    deliveredAt: Optional[datetime] = None  # Delivery timestamp
    notes: Optional[str] = None  # Special notes/instructions
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "orders"
        indexes = [
            [("user", 1)],  # Index for quick user order lookup
            [("created_at", -1)],  # Index for sorting by date
            [("orderStatus", 1)]  # Index for status filtering
        ]

File: backend/models/Order.js
const mongoose = require('mongoose');

const orderSchema = new mongoose.Schema({
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    orderItems: [{
        product: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
        name: String,
        image: String,
        price: Number,
        quantity: { type: Number, required: true, min: 1 }
    }],
    shippingAddress: {
        fullName: { type: String, required: true },
        phone: { type: String, required: true },
        street: { type: String, required: true },
        city: { type: String, required: true },
        state: { type: String, required: true },
        zipCode: { type: String, required: true },
        country: { type: String, default: 'India' }
    },
    paymentMethod: { type: String, required: true, enum: ['COD', 'Online'], default: 'COD' },
    paymentResult: {
        id: String,
        status: String,
        updateTime: String,
        emailAddress: String
    },
    itemsPrice: { type: Number, required: true, default: 0 },
    taxPrice: { type: Number, required: true, default: 0 },
    shippingPrice: { type: Number, required: true, default: 0 },
    discountAmount: { type: Number, default: 0 },
    totalPrice: { type: Number, required: true, default: 0 },
    couponCode: { type: String, default: '' },
    isPaid: { type: Boolean, default: false },
    paidAt: Date,
    status: {
        type: String,
        enum: ['Processing', 'Confirmed', 'Shipped', 'Delivered', 'Cancelled'],
        default: 'Processing'
    },
    isDelivered: { type: Boolean, default: false },
    deliveredAt: Date
}, { timestamps: true });

module.exports = mongoose.model('Order', orderSchema);
